Add stickiness support for NLB target_groups

madpipeline opened the pull request:
<!--- See what makes a good Pull Request at : https://github.com/terraform-providers/terraform-provider-aws/blob/master/docs/CONTRIBUTING.md --->

<!--- Please keep this note for the community --->

### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

<!--- Thank you for keeping this note for the community --->

<!--- If your PR fully resolves and should automatically close the linked issue, use Closes. Otherwise, use Relates --->
Closes #9093 
Closes #13762

Release note for [CHANGELOG](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md):

```release-note
Allow `stickiness` to be set for NLB `aws_lb_target_group` resources. 
Allow `source_ip` as only valid `stickiness` `type` for `aws_lb_target_group` resource.
```

This PR is meant to update the #13762, that has a minor syntax issue and no response from the author.

YakDriver reviewed (changes requested):
@madpipeline This looks great! Thank you for your work on this. Acceptance tests are passing. There are only two things we need before we can merge: a documentation update (`website/docs/r/lb_target_group.html.markdown`) and a test for the new value (`aws/resource_aws_lb_target_group_test.go`). Would you be willing to add these?

```
--- PASS: TestLBTargetGroupCloudwatchSuffixFromARN (0.54s)
--- PASS: TestAccAWSLBTargetGroup_withoutHealthcheck (49.26s)
--- PASS: TestAccAWSLBTargetGroup_Protocol_Tls (55.92s)
--- PASS: TestAccAWSLBTargetGroup_BackwardsCompatibility (64.22s)
--- PASS: TestAccAWSLBTargetGroup_basicUdp (64.93s)
--- PASS: TestAccAWSLBTargetGroup_namePrefix (64.95s)
--- PASS: TestAccAWSLBTargetGroup_generatedName (67.61s)
--- PASS: TestAccAWSLBTargetGroup_basic (81.96s)
--- PASS: TestAccAWSLBTargetGroupAttachment_lambda (83.06s)
--- PASS: TestAccAWSLBTargetGroup_networkLB_TargetGroupWithProxy (113.30s)
--- PASS: TestAccAWSLBTargetGroup_changeNameForceNew (120.44s)
--- PASS: TestAccAWSLBTargetGroup_TCP_HTTPHealthCheck (123.12s)
--- PASS: TestAccAWSLBTargetGroup_stickinessWithTCPEnabledShouldError (11.54s)
--- PASS: TestAccAWSLBTargetGroup_changeProtocolForceNew (126.08s)
--- PASS: TestAccAWSLBTargetGroup_changePortForceNew (126.42s)
--- PASS: TestAccAWSLBTargetGroup_Protocol_Tcp_HealthCheck_Protocol (127.86s)
--- PASS: TestAccAWSLBTargetGroupAttachment_BackwardsCompatibility (129.86s)
--- PASS: TestAccAWSLBTargetGroup_defaults_application (63.61s)
--- PASS: TestAccAWSLBTargetGroupAttachment_basic (133.31s)
--- PASS: TestAccAWSLBTargetGroup_stickinessWithTCPDisabled (52.81s)
--- PASS: TestAccAWSLBTargetGroupAttachment_Port (140.71s)
--- PASS: TestAccAWSLBTargetGroupAttachment_disappears (143.58s)
--- PASS: TestAccAWSLBTargetGroup_changeVpcForceNew (97.31s)
--- PASS: TestAccAWSLBTargetGroup_defaults_network (65.93s)
--- PASS: TestAccAWSLBTargetGroup_enableHealthCheck (85.06s)
--- PASS: TestAccAWSLBTargetGroupAttachment_ipAddress (150.13s)
--- PASS: TestAccAWSLBTargetGroup_updateHealthCheck (87.33s)
--- PASS: TestAccAWSLBTargetGroup_networkLB_TargetGroup (152.79s)
--- PASS: TestAccAWSLBTargetGroup_tags (105.88s)
--- PASS: TestAccAWSLBTargetGroup_updateSticknessEnabled (99.57s)
```

madpipeline commented:
Hey @YakDriver. Sure. I can handle the docs, but I'm not sure how to add the test, since I see right now there are no UDP tests, and I'd like to keep your tests pattern. Any advice on how to add that test? I could use a hand with it.

YakDriver commented:
No problem. The 10,000 ft overview is that the basic test should test everything possible. Each additional test should only test specifically what it's about since the basic test has the bases covered. Here's a [short overview](https://medium.com/faun/terraform-acceptance-testing-the-basics-5450d35a4421) of acceptance testing.

For context, the basic test for `aws_lb_target_group` is here: https://github.com/terraform-providers/terraform-provider-aws/blob/baf451417714e7e3498b5468a7bd39727c647ca5/aws/resource_aws_lb_target_group_test.go#L94

You want to create a test and the test will need config. This will involve adding probably two functions: the test and a function that returns a big string with the config/HCL. You can use this test as a template: https://github.com/terraform-providers/terraform-provider-aws/blob/baf451417714e7e3498b5468a7bd39727c647ca5/aws/resource_aws_lb_target_group_test.go#L967-L989. The config for that test is here: https://github.com/terraform-providers/terraform-provider-aws/blob/baf451417714e7e3498b5468a7bd39727c647ca5/aws/resource_aws_lb_target_group_test.go#L1926-L1948.

Basically, copy those functions, rename them, change the config, make sure the test is testing for what you need it to, and Bob's your uncle.

Here are [more in-depth docs](https://github.com/terraform-providers/terraform-provider-aws/blob/master/docs/contributing/running-and-writing-acceptance-tests.md) and we're always here to help!

madpipeline commented:
It seems that the changes are not so easy. Stickiness is now allowed with NLBs acording to [the docs](https://docs.aws.amazon.com/elasticloadbalancing/latest/network/load-balancer-target-groups.html#target-group-attributes). So I'll make some more changes to alter the TCP behaviour regarding stickiness.

Not sure yet if embarking on this change wagon is a good idea or not at the current time.

madpipeline commented:
I have a WIP for the tests, but it's not clean yet, so I'm not ready to commit them.

YakDriver commented:
@madpipeline Excellent! Thanks for continuing on with this. I recognize this is probably not the only thing you've got going on so I appreciate your willingness to continue and keeping us informed about your status. If you get stuck or want to bounce anything off me, please feel free. (I add the "waiting response" label to help organize things on my end and not to put any pressure on you!)

YakDriver commented:
@madpipeline We are hoping to get this merged by Oct 14, for inclusion in the 3.11 provider release. What's your schedule like?

madpipeline commented:
Code is done. Let's see what the CI gods have to say about this.

madpipeline commented:
Tests seem ok. However, I can't find in the CI logs the per test output you've sent in a previous message. I wanted to see the new tests I've added in that list.

YakDriver commented:
@madpipeline My first impression is that this is looking great! Thanks for your work on it. I have some quick questions and will take a closer look shortly.

* I notice that you've added a `stickiness` block to many of the tests that didn't have them before. Is this necessary? If they didn't need one before, it seems like they should not need one now. If you're adding them simply to test the functionality, adding the block to one existing test, in addition to your new tests, should be sufficient.
* Does the API have enums for the `lb_cookie` and `source_ip` literals? If so, we should use those.

madpipeline commented:
@YakDriver I've added those before I've figured out how to set the default type for NLB TG. They are not needed. I've removed them.

I wanted to use API enums for the type literals, but didn't find any. I've considered creating my own `const`s but they are not used that many times in each fine, so I gave up on that idea.

YakDriver reviewed (approved):
@madpipeline Thank you for all your help with this! I made a few changes in order to get it in. I share with you a couple of lessons from this:

1. The creation process usually involves 3 parts: Create, Update, Read. In order to reduce duplication in the code, generally Create does a very basic create of a skeleton resource. The Create function only needs to check for things that would prevent the creation of the "skeleton". Then, the creation process goes to Update where everything is modified to look like the config. Most of the work is done in Update. Finally, the last part of the creation process is Read. This basically checks to make sure the config, state, and infrastructure are the same. If not, you get a non-empty plan.
2. Generally you want to avoid breaking changes. For example, there is a default `cookie_duration`. This could cause problems with non-empty plans. The config says 86400 even when there is no `cookie_duration` such as should be the case for an NLB. Getting rid of the default could break people's configs if they rely on it. Thus, I kept the default but added a `DiffSuppressFunc` to avoid the default causing problems.
3. If possible, let the AWS API do the error checking. It does take longer to wait for errors to come back but it is more sustainable. The API changes a lot and if we duplicate the error checking, we'll have to change a lot too. Here, instead of checking for problems with stickiness types and protocols, I let the documentation and the AWS API take care of it. I was able to remove all the error checking for that.

Let me know if you have questions! Thank you for all your work on this. Great job! 💯

YakDriver commented:
The output from acceptance tests (Commercial):

```
--- PASS: TestAccAWSLBTargetGroup_withoutHealthcheck (19.98s)
--- PASS: TestAccAWSLBTargetGroup_stickinessDefaultALB (35.46s)
--- PASS: TestAccAWSLBTargetGroup_namePrefix (35.54s)
--- PASS: TestAccAWSLBTargetGroup_defaults_network (43.91s)
--- PASS: TestAccAWSLBTargetGroup_BackwardsCompatibility (45.35s)
--- PASS: TestAccAWSLBTargetGroup_defaults_application (47.32s)
--- PASS: TestAccAWSLBTargetGroup_generatedName (48.84s)
--- PASS: TestAccAWSLBTargetGroup_stickinessInvalidALB (54.16s)
--- PASS: TestAccAWSLBTargetGroup_stickinessInvalidNLB (57.75s)
--- PASS: TestAccAWSLBTargetGroup_basic (59.56s)
--- PASS: TestAccAWSLBTargetGroup_Protocol_Tls (26.88s)
--- PASS: TestAccAWSLBTargetGroup_changePortForceNew (65.38s)
--- PASS: TestAccAWSLBTargetGroup_changeNameForceNew (66.73s)
--- PASS: TestAccAWSLBTargetGroup_stickinessValidALB (66.98s)
--- PASS: TestAccAWSLBTargetGroup_TCP_HTTPHealthCheck (69.98s)
--- PASS: TestAccAWSLBTargetGroup_changeProtocolForceNew (70.97s)
--- PASS: TestAccAWSLBTargetGroup_changeVpcForceNew (74.88s)
--- PASS: TestAccAWSLBTargetGroup_basicUdp (30.61s)
--- PASS: TestAccAWSLBTargetGroup_enableHealthCheck (32.60s)
--- PASS: TestAccAWSLBTargetGroup_Protocol_Tcp_HealthCheck_Protocol (66.01s)
--- PASS: TestAccAWSLBTargetGroup_updateSticknessEnabled (91.06s)
--- PASS: TestAccAWSLBTargetGroup_stickinessDefaultNLB (92.93s)
--- PASS: TestAccAWSLBTargetGroup_networkLB_TargetGroupWithProxy (58.09s)
--- PASS: TestAccAWSLBTargetGroup_updateHealthCheck (64.01s)
--- PASS: TestAccAWSLBTargetGroup_networkLB_TargetGroup (112.27s)
--- PASS: TestAccAWSLBTargetGroup_stickinessValidNLB (114.47s)
--- PASS: TestAccAWSLBTargetGroup_tags (81.13s)
```

The output from acceptance tests (GovCloud):

```
--- PASS: TestAccAWSLBTargetGroup_stickinessDefaultALB (39.73s)
--- PASS: TestAccAWSLBTargetGroup_generatedName (39.78s)
--- PASS: TestAccAWSLBTargetGroup_BackwardsCompatibility (45.48s)
--- PASS: TestAccAWSLBTargetGroup_enableHealthCheck (45.79s)
--- PASS: TestAccAWSLBTargetGroup_defaults_network (48.42s)
--- PASS: TestAccAWSLBTargetGroup_basic (53.77s)
--- PASS: TestAccAWSLBTargetGroup_defaults_application (56.01s)
--- PASS: TestAccAWSLBTargetGroup_stickinessInvalidALB (62.96s)
--- PASS: TestAccAWSLBTargetGroup_stickinessInvalidNLB (63.08s)
--- PASS: TestAccAWSLBTargetGroup_withoutHealthcheck (22.01s)
--- PASS: TestAccAWSLBTargetGroup_Protocol_Tls (29.95s)
--- PASS: TestAccAWSLBTargetGroup_stickinessValidALB (72.45s)
--- PASS: TestAccAWSLBTargetGroup_namePrefix (35.46s)
--- PASS: TestAccAWSLBTargetGroup_networkLB_TargetGroupWithProxy (77.26s)
--- PASS: TestAccAWSLBTargetGroup_changePortForceNew (79.60s)
--- PASS: TestAccAWSLBTargetGroup_TCP_HTTPHealthCheck (80.24s)
--- PASS: TestAccAWSLBTargetGroup_changeNameForceNew (82.10s)
--- PASS: TestAccAWSLBTargetGroup_Protocol_Tcp_HealthCheck_Protocol (83.67s)
--- PASS: TestAccAWSLBTargetGroup_changeVpcForceNew (84.85s)
--- PASS: TestAccAWSLBTargetGroup_changeProtocolForceNew (88.29s)
--- PASS: TestAccAWSLBTargetGroup_basicUdp (35.55s)
--- PASS: TestAccAWSLBTargetGroup_stickinessDefaultNLB (103.41s)
--- PASS: TestAccAWSLBTargetGroup_tags (116.82s)
--- PASS: TestAccAWSLBTargetGroup_updateHealthCheck (70.59s)
--- PASS: TestAccAWSLBTargetGroup_stickinessValidNLB (133.45s)
--- PASS: TestAccAWSLBTargetGroup_updateSticknessEnabled (94.76s)
--- PASS: TestAccAWSLBTargetGroup_networkLB_TargetGroup (114.37s)
```

madpipeline commented on aws/resource_aws_lb_target_group.go:
```diff
@@ -282,39 +289,9 @@ func resourceAwsLbTargetGroupCreate(d *schema.ResourceData, meta interface{}) er
-				}
-			}
-		} else {
-			if d.Get("protocol").(string) == elbv2.ProtocolEnumTcp ||
```
By removing this `else` block, how are we ensuring that if the `stickiness` block isn't defined, the right type is used?

madpipeline replied:
I expect the `TestAccAWSLBTargetGroup_stickinessDefaultNLB` test to fail.

madpipeline replied:
I see it passing, but I don't understand why.

YakDriver replied:
The API will set the stickiness and set enabled to false.

madpipeline commented on aws/resource_aws_lb_target_group.go:
```diff
@@ -158,6 +158,13 @@ func resourceAwsLbTargetGroup() *schema.Resource {
 							Optional:     true,
 							Default:      86400,
 							ValidateFunc: validation.IntBetween(0, 604800),
+							DiffSuppressFunc: func(k, old, new string, d *schema.ResourceData) bool {
```
Cool, didn't know about this.

YakDriver replied:
It may not be the best way to do this but I did not want to remove the default. At least it's easy to understand.

YakDriver commented:
I'm guessing that as people use this, other problems will surface. But, that's the nature of small incremental changes made often! Thanks again for your contribution. We look forward to your help again!

YakDriver commented:
Potentially help notes for this resource that I'll leave here:

```
ALB = HTTP, HTTPS
deregistration_delay.timeout_seconds
stickiness.enabled
stickiness.type = lb_cookie
load_balancing.algorithm.type
slow_start.duration_seconds
stickiness.lb_cookie.duration_seconds

NLB = TCP, UDP, TCP_UDP
deregistration_delay.timeout_seconds
stickiness.enabled
stickiness.type = source_ip
proxy_protocol_v2.enabled

NLB = TLS
deregistration_delay.timeout_seconds
stickiness.enabled = false
stickiness.type = source_ip
proxy_protocol_v2.enabled
```

YakDriver commented:
Closes #10494

madpipeline commented:
> Potentially helpful notes for this resource that I'll leave here:
> 
> ```
> ALB = HTTP, HTTPS
> deregistration_delay.timeout_seconds
> stickiness.enabled
> stickiness.type = lb_cookie
> load_balancing.algorithm.type
> slow_start.duration_seconds
> stickiness.lb_cookie.duration_seconds
> 
> NLB = TCP, UDP, TCP_UDP
> deregistration_delay.timeout_seconds
> stickiness.enabled
> stickiness.type = source_ip
> proxy_protocol_v2.enabled
> 
> NLB = TLS
> deregistration_delay.timeout_seconds
> stickiness.enabled = false
> stickiness.type = source_ip
> proxy_protocol_v2.enabled
> ```

These should go in a new issue. It will get buried here.
